oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,142 @@
+
+    public Downloader(int concurrency) {
+        LOG.info("Initializing Downloader with max number of concurrent requests={}", concurrency);
+        this.executorService = new ThreadPoolExecutor(
+                (int) Math.ceil(concurrency * .1), concurrency, 60L, TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("downloader-%d")
+                        .setDaemon(true)
+                        .build()
+        );
+    }
+
+    public List<ItemResponse> download(List<Item> items) {
+        LOG.debug("Preparing to download {} items.\n{}", items.size(), items);
+        try {
+            return executorService
+                    .invokeAll(items.stream().map(DownloadWorker::new).collect(Collectors.toList()))
+                    .stream()
+                    .map(itemResponseFuture -> {
+                        try {
+                            return itemResponseFuture.get();
+                        } catch (InterruptedException | ExecutionException e) {
+                            throw new RuntimeException(e);
+                        }
+                    })
+                    .collect(Collectors.toList());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        executorService.shutdown();
+    }
+
+    public static class Item {
+        public String source;
+        public String destination;
+
+        @Override
+        public String toString() {
+            return "Item{" +
+                    "source='" + source + '\'' +
+                    ", destination='" + destination + '\'' +
+                    '}';
+        }
+    }
+
+    public static class ItemResponse {
+        public final Item item;
+        public boolean failed;
+        public long size;
+        public long time;
+        public Throwable throwable;
+
+        public ItemResponse(Item item) {
+            this.item = item;
+        }
+    }
+
+    private static class DownloadWorker implements Callable<ItemResponse> {
+
+        private final Item item;
+
+        DownloadWorker(Item item) {
+            this.item = item;
+        }
+
+        @Override
+        public ItemResponse call() {
+            ItemResponse response  = new ItemResponse(item);
+            long t0 = System.currentTimeMillis();
+            try {
+                URL sourceUrl = new URL(item.source);
+                File destinationFile = new File(item.destination);
+                destinationFile.getParentFile().mkdirs();
+                try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.openStream());
+                     FileOutputStream outputStream = new FileOutputStream(destinationFile)) {
+                    response.size = outputStream.getChannel()
+                            .transferFrom(byteChannel, 0, Long.MAX_VALUE);
```

Comment:
I don't think this will do a direct copy. The `byteChannel` here is an instance of `ReadableByteChannelImpl` (created by `Channels.newChannel`), which takes a generic InputStream, so has no knowledge of the implementation details of this input stream. The `transferFrom` method cannot guess that this inputstream is backed by a socket, so it will likely revert to copying the data using a buffer in the JVM heap. Here is the  source code for `FileChannelImpl#transferFrom` (in JDK 19):

```
        if (src instanceof FileChannelImpl fci) {
            long n;
            if ((n = transferFromDirectly(fci, position, count)) >= 0)
                return n;
            if ((n = transferFromFileChannel(fci, position, count)) >= 0)
                return n;
        }

        return transferFromArbitraryChannel(src, position, count);
```

 So it will fall to `transferFromArbitraryChannel`, which just does a normal copy. 

I found examples of doing a direct copy from a SocketChannel, but I'm not sure if it is possible to open a SocketChannel from an URL connection.

Otherwise, if we do not use direct copy, it may be better to wrap the input and output streams in buffers, that may speed up the download.

Reply:
If I'm not mistaken (but I didn't check), transferFromArbitraryChannel does already use buffers... so I see the usage of transferFrom to be a convenience... But I might be wrong.

Reply:
Yes, I had a look at the source code of `transferFromArbitraryChannel` and it does use an internal buffer of 8KB. So you are right, adding more buffering is not needed and would likely reduce performance.

Reply:
Direct copy does not happen with a ReadableFileChannel coming from an URL. As you already noticed, `transferFromArbitraryChannel` uses an internal buffer. From the first tests, the results are acceptable.

It seems that the apache HTTP client has some utility to perform a zero-copy download. We can further improve this in the future if needed.